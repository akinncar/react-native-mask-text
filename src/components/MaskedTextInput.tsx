import React, {
  useEffect,
  useState,
  forwardRef,
  ForwardRefRenderFunction,
} from 'react'
import { TextInput, TextInputProps } from 'react-native'
import { mask, unMask } from '../utils/mask'
import type { FormatType, MaskOptions, StyleObj, TextDecorationOptions } from 'src/@types'


type TIProps = Omit<TextInputProps, 'onChangeText'>
export interface MaskedTextInputProps extends TIProps {
  mask?: string
  type?: FormatType
  options?: MaskOptions
  defaultValue?: string
  onChangeText: (text: string, rawText: string) => void
  inputAccessoryView?: JSX.Element
  textBold?: boolean
  textItalic?:boolean
  textDecoration?:TextDecorationOptions
  style?: StyleObj
}

export const MaskedTextInputComponent: ForwardRefRenderFunction<
  TextInput,
  MaskedTextInputProps
> = (
  {
    mask: pattern = '',
    type = 'custom',
    options = {} as MaskOptions,
    defaultValue,
    onChangeText,
    value,
    inputAccessoryView,
    autoCapitalize = 'sentences',
    textBold,
    textItalic,
    textDecoration,
    style,
    ...rest
  },
  ref
): JSX.Element => {
  const styleSheet = [
    {
      fontWeight: textBold && 'bold',
      fontStyle: textItalic && 'italic',
      textDecorationLine: textDecoration
    },
    style
  ]
  const getMaskedValue = (value: string) =>
    mask(value, pattern, type, options, autoCapitalize)
  const getUnMaskedValue = (value: string) =>
    unMask(value, type as 'custom' | 'currency')

  const defaultValueCustom = defaultValue || ''
  const defaultValueCurrency = defaultValue || '0'

  const initialRawValue = value;

  const initialMaskedValue = getMaskedValue(
    type === 'currency' ? defaultValueCurrency : defaultValueCustom
  )

  const initialUnMaskedValue = getUnMaskedValue(
    type === 'currency' ? defaultValueCurrency : defaultValueCustom
  )

  const [maskedValue, setMaskedValue] = useState(initialMaskedValue)
  const [unMaskedValue, setUnmaskedValue] = useState(initialUnMaskedValue)
  const [rawValue, setRawValue] = useState(initialRawValue);
  const [isInitialRender, setIsInitialRender] = useState(true)

  const actualValue = pattern || type === "currency" ? maskedValue : rawValue;

  function onChange(value: string) {
    const newUnMaskedValue = unMask(value, type as 'custom' | 'currency')
    const newMaskedValue = mask(newUnMaskedValue, pattern, type, options)

    setMaskedValue(newMaskedValue)
    setUnmaskedValue(newUnMaskedValue)
    setRawValue(value);
  }

  useEffect(() => {
    if (isInitialRender) {
      setIsInitialRender(false)
      return
    }

    onChangeText(maskedValue, unMaskedValue)
  }, [maskedValue, unMaskedValue])

  useEffect(() => {
    if (value) {
      setMaskedValue(getMaskedValue(value))
      setUnmaskedValue(getUnMaskedValue(value))
    } else {
      setMaskedValue(initialMaskedValue)
      setUnmaskedValue(initialUnMaskedValue)
    }
  }, [value])

  return (
    <>
      <TextInput
        onChangeText={(value) => onChange(value)}
        ref={ref}
        maxLength={pattern.length || undefined}
        autoCapitalize={autoCapitalize}
        {...rest}
        value={actualValue}
        style={styleSheet as StyleObj}
      />
      {inputAccessoryView}
    </>
  )
}

export const MaskedTextInput = forwardRef(MaskedTextInputComponent)
