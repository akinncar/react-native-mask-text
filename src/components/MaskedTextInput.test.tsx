import React from 'react'
import { render, fireEvent, waitFor } from '@testing-library/react-native';
import { MaskedTextInput } from './MaskedTextInput';
import { Button, Keyboard, InputAccessoryView } from 'react-native';


describe('<MaskedTextInput />', () => {
  const mockedOnChangeText = jest.fn();

  test('should render correctly without a mask', () => {
    const container = render(
      <MaskedTextInput value="churrasco" onChangeText={mockedOnChangeText} />,
    );
    expect(container).toMatchSnapshot();
  })

  test('should renders correctly with custom mask', () => {
    const container = render(
      <MaskedTextInput mask="AAA-999" onChangeText={mockedOnChangeText} />,
    );
    expect(container).toMatchSnapshot();
  });

  test('should renders correctly with custom mask default value', () => {
    const container = render(
      <MaskedTextInput
        mask="AAA-999"
        onChangeText={mockedOnChangeText}
        defaultValue="ABC-123"
      />
    );
    expect(container.getByDisplayValue('ABC-123')).toBeTruthy()
  });

  test('should renders correctly with currency mask', () => {
    const container = render(
      <MaskedTextInput
        type="currency"
        options={{
          prefix: '$',
          decimalSeparator: '.',
          groupSeparator: ',',
          precision: 2,
        }}
        onChangeText={mockedOnChangeText}
      />
    );
    expect(container).toMatchSnapshot();
  });

  test('should mask input text with custom mask', async () => {
    const container = render(
      <MaskedTextInput
        mask="AAA-999"
        onChangeText={mockedOnChangeText}
        testID="masked-text-input"
      />
    );

    fireEvent.changeText(container.getByTestId('masked-text-input'), 'RCT777')

    await waitFor(() => {
      expect(container.getByDisplayValue('RCT-777')).toBeTruthy()
    })
  });

  test('should mask input text with currency mask', async () => {
    const container = render(
      <MaskedTextInput
        type="currency"
        options={{
          prefix: '$',
          decimalSeparator: '.',
          groupSeparator: ',',
          precision: 2,
        }}
        onChangeText={mockedOnChangeText}
        testID="masked-text-input"
      />
    );

    fireEvent.changeText(container.getByTestId('masked-text-input'), '5999')

    await waitFor(() => {
      expect(container.getByDisplayValue('$59.99')).toBeTruthy()
    })
  });

  test('should renders correctly with an accessory view', () => {
    const container = render(
      <MaskedTextInput         
      type="currency"
      options={{
        prefix: '$',
        decimalSeparator: '.',
        groupSeparator: ',',
        precision: 2,
      }} 
      onChangeText={mockedOnChangeText} 
      inputAccessoryViewID='Done'
      inputAccessoryView={
        <InputAccessoryView nativeID='Done'>
        <Button
          onPress={() => Keyboard.dismiss()}
          title="Clear text"
        />
        </InputAccessoryView>
      }
      />,
    );

    const tree = container.toJSON();
    expect(tree[0].props.inputAccessoryViewID).toBe('Done');
    expect(tree[1].props.nativeID).toBe('Done');
  });
  test('should be bold when the textBold attribute is added', () => {
    const container  =
    render(<MaskedTextInput 
      testID="masked-text-input" 
      onChangeText={mockedOnChangeText} 
      textBold 
      mask="99/99/9999"
      />)
      expect(container.getByTestId('masked-text-input')).toHaveStyle({fontWeight: 'bold' });
    });
    test('should be italic when the textItalic attribute is added', () => {
      const container  =
      render(<MaskedTextInput 
        testID="masked-text-input" 
        onChangeText={mockedOnChangeText} 
        textItalic 
        mask="99/99/9999"
        />)
        expect(container.getByTestId('masked-text-input')).toHaveStyle({fontStyle: 'italic' });
    });
    test('should be the line style added to the text when the textDecoration attribute has a non-null value', () => {
      const textDecorationLine = 'underline'
      const container  =
      render(<MaskedTextInput
        testID="masked-text-input" 
        onChangeText={mockedOnChangeText} 
        textDecoration={textDecorationLine} 
        mask="99/99/9999"
        />)
        expect(container.getByTestId('masked-text-input')).toHaveStyle({textDecorationLine: textDecorationLine });
    });  
});
