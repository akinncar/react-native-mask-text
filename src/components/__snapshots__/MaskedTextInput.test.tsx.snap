// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MaskedTextInput /> should render correctly without a mask 1`] = `
<TextInput
  allowFontScaling={true}
  onChangeText={[Function]}
  rejectResponderTermination={true}
  style={
    Object {
      "fontStyle": undefined,
      "fontWeight": undefined,
      "textDecorationLine": undefined,
    }
  }
  underlineColorAndroid="transparent"
  value="with space and special* characters;"
/>
`;

exports[`<MaskedTextInput /> should renders correctly with currency mask 1`] = `
<TextInput
  allowFontScaling={true}
  onChangeText={[Function]}
  rejectResponderTermination={true}
  style={
    Object {
      "fontStyle": undefined,
      "fontWeight": undefined,
      "textDecorationLine": undefined,
    }
  }
  underlineColorAndroid="transparent"
  value="$0.00"
/>
`;

exports[`<MaskedTextInput /> should renders correctly with custom mask 1`] = `
<TextInput
  allowFontScaling={true}
  maxLength={7}
  onChangeText={[Function]}
  rejectResponderTermination={true}
  style={
    Object {
      "fontStyle": undefined,
      "fontWeight": undefined,
      "textDecorationLine": undefined,
    }
  }
  underlineColorAndroid="transparent"
  value=""
/>
`;
